python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,836 @@
+
+            # The hash that acts as the key we'll use to look up our cached
+            # data will be based on the entity type and the pipeline config's
+            # descriptor uri. We can get the descriptor from the toolkit
+            # manager and pass that through along with the entity type from SG
+            # to the core hook that computes the hash.
+            manager.pipeline_configuration = pc["id"]
+            pc_descriptor = manager.resolve_descriptor(project_entity)
+
+            self.logger.debug("Resolved config descriptor: %r" % pc_descriptor)
+            pc_key = pc_descriptor.get_uri()
+
+            pc_data = dict()
+            pc_data["contents_hash"] = self._get_contents_hash(
+                pc_descriptor,
+                self._get_site_state_data(),
+            )
+            pc_data["lookup_hash"] = self._get_lookup_hash(
+                pc_key,
+                project_entity,
+                data["entity_type"],
+            )
+            pc_data["descriptor"] = pc_descriptor
+            pc_data["entity"] = pc
+            config_data[pc["id"]] = pc_data
+
+        with self._db_connect() as (connection, cursor):
+            for pc in pcs:
+                # The hash that acts as the key we'll use to look up our cached
+                # data will be based on the entity type and the pipeline config's
+                # descriptor uri. We can get the descriptor from the toolkit
+                # manager and pass that through along with the entity type from SG
+                # to the core hook that computes the hash.
+                pc_data = config_data[pc["id"]]
+                manager.pipeline_configuration = pc["id"]
+                pc_descriptor = pc_data["descriptor"]
+                lookup_hash = pc_data["lookup_hash"]
+                contents_hash = pc_data["contents_hash"]
+
+                cursor.execute(
+                    "SELECT commands, contents_hash FROM engine_commands WHERE lookup_hash=?",
+                    (lookup_hash,)
+                )
+                cached_data = list(cursor.fetchone() or [])
+
+                try:
+                    if cached_data:
+                        cached_contents_hash = cached_data[1]
+
+                        if str(contents_hash) == str(cached_contents_hash):
+                            self.logger.debug("Cache is up to date.")
+                            actions = self._process_commands(
+                                commands=cPickle.loads(str(cached_data[0])),
+                                project=project_entity,
+                                entities=entities,
+                            )
+                            self.logger.debug("Actions found in cache: %s" % actions)
+                            all_actions[pc["name"]] = dict(
+                                actions=self._filter_by_project(
+                                    actions,
+                                    self._get_software_entities(),
```

Comment:
this feels like it should be done in the cache process? if we just use the sw entity as part of the invalidation key, then we could capture this at cache time instead perhaps? i guess i am wondering if we can move as much of the 'compute' as possible into the caching part and all that the main thread is doing is basically figuring out 'am i still up to date' and if the anwer is no, then fire up the cache script that does all the computation